src/thread/queue.cr

```diff
@@ -0,0 +1,53 @@
+# Based on rubysl-thread ruby gem implementation
+# Copyright (c) 2013, Brian Shirai
+# Licensed under the BSD 3-clause
+
+# :nodoc:
+class Thread(T, R)
+  # :nodoc:
+  class Queue(T)
+    class Error < Exception; end
+
+    def initialize
+      @que = Deque(T).new(16)
+      @mutex = Thread::Mutex.new
+      @resource = Thread::ConditionVariable.new
+    end
+
+    def push(item : T)
+      @mutex.synchronize do
+        @que.push(item)
+        @resource.signal
+      end
+    end
+
+    def pop(blocking = true)
+      loop do
+        @mutex.synchronize do
+          if @que.empty?
+            raise Error.new("queue is empty") unless blocking
```

Comment:
When called from the event-loop, this should always non blocking, otherwise the next call will block the event loop until a result a available (bad).

It could also merely return nil, allowing the calling Fiber to yield itself, instead of spending resources raising an exception, catching it and then yielding. Thought there won't be any mean to differentiate between a "would block" and a "popping nil" scenarios.